Data node v2 internal error when asked about missing proposal

# Problem encountered
Query the data node v2 proposal API for a proposal that cannot be found.

# Observed behaviour
RPC internal error

# Expected behaviour
V1 behaviour:
        # Expect to not find the proposal.
>       governMod.GetProposalByID(market_id, expect_err='"grpc_message":"NotFound error","grpc_status":5')

## Automation
Use branch issue/516_extended until [this PR](https://github.com/vegaprotocol/system-tests/pull/762/files) is merged
pytest -k test_proposed_market_with_votes_not_restored

# Evidence
E           grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
E               status = StatusCode.INTERNAL
E               details = "Internal error"
E               debug_error_string = "{"created":"@1651761737.677227900","description":"Error received from peer ipv4:127.0.0.1:3027","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Internal error","grpc_status":13}"
E           >

/home/jake/.local/lib/python3.8/site-packages/grpc/_channel.py:826: _InactiveRpcError

## Logs
[logs.txt](https://github.com/vegaprotocol/data-node/files/8632260/logs.txt)

## Additional context
Add any other context about the problem here including; system version numbers, components affected.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,7 @@
 
 ### 🚨 Breaking changes
 - [518](https://github.com/vegaprotocol/data-node/issues/518) - Free-form properties are moved to rationale.
-- 
+-
 ### 🗑️  Deprecation
 - [](https://github.com/vegaprotocol/data-node/issues/xxx) -
 
@@ -17,12 +17,13 @@
 - [518](https://github.com/vegaprotocol/data-node/issues/518) - Add rationale to proposals
 - [526](https://github.com/vegaprotocol/data-node/issues/526) - Add market id and reward type to reward and market to transfer
 - [540](https://github.com/vegaprotocol/data-node/issues/540) - CI: trigger Devnet deployment on merges to develop branch
-- [546](https://github.com/vegaprotocol/data-node/issues/546) - Data retention for margin levels 
+- [546](https://github.com/vegaprotocol/data-node/issues/546) - Data retention for margin levels
 - [553](https://github.com/vegaprotocol/data-node/issues/553) - Update transfers API to expose dispatch strategy
 - [578](https://github.com/vegaprotocol/data-node/issues/578) - Add metrics for `SQL` queries
 - [582](https://github.com/vegaprotocol/data-node/issues/582) - Add a cache for assets
-- [548](https://github.com/vegaprotocol/data-node/issues/548) - Remove foreign key constraints on hyper tables 
-- [591](https://github.com/vegaprotocol/data-node/issues/591) - Optimise liquidity provision and margin levels data retention and storage 
+- [548](https://github.com/vegaprotocol/data-node/issues/548) - Remove foreign key constraints on hyper tables
+- [591](https://github.com/vegaprotocol/data-node/issues/591) - Optimise liquidity provision and margin levels data retention and storage
+- [588](https://github.com/vegaprotocol/data-node/issues/588) - Return correct error code when proposal not found
 
 ### 🐛 Fixes
 - [524](https://github.com/vegaprotocol/data-node/issues/524) - Fix for incorrect balances

--- a/api/trading_data_delegator.go
+++ b/api/trading_data_delegator.go
@@ -421,7 +421,9 @@ func (t *tradingDataDelegator) GetProposalByID(ctx context.Context,
 	defer metrics.StartAPIRequestAndTimeGRPC("GetProposalByID SQL")()
 
 	proposal, err := t.proposalsStore.GetByID(ctx, req.ProposalId)
-	if err != nil {
+	if errors.Is(err, sqlstore.ErrProposalNotFound) {
+		return nil, apiError(codes.NotFound, ErrMissingProposalID, err)
+	} else if err != nil {
 		return nil, apiError(codes.Internal, ErrNotMapped, err)
 	}
 
@@ -439,7 +441,9 @@ func (t *tradingDataDelegator) GetProposalByReference(ctx context.Context,
 	defer metrics.StartAPIRequestAndTimeGRPC("GetProposalByID SQL")()
 
 	proposal, err := t.proposalsStore.GetByReference(ctx, req.Reference)
-	if err != nil {
+	if errors.Is(err, sqlstore.ErrProposalNotFound) {
+		return nil, apiError(codes.NotFound, ErrMissingProposalReference, err)
+	} else if err != nil {
 		return nil, apiError(codes.Internal, ErrNotMapped, err)
 	}
 

--- a/sqlstore/proposals.go
+++ b/sqlstore/proposals.go
@@ -8,7 +8,10 @@ import (
 	"code.vegaprotocol.io/data-node/entities"
 	"code.vegaprotocol.io/data-node/metrics"
 	"github.com/georgysavva/scany/pgxscan"
+	"github.com/pkg/errors"
 )
+
+var ErrProposalNotFound = errors.New("proposal not found")
 
 type Proposals struct {
 	*ConnectionSource
@@ -56,6 +59,10 @@ func (ps *Proposals) GetByID(ctx context.Context, id string) (entities.Proposal,
 	var p entities.Proposal
 	query := `SELECT * FROM proposals_current WHERE id=$1`
 	err := pgxscan.Get(ctx, ps.Connection, &p, query, entities.NewProposalID(id))
+	if pgxscan.NotFound(err) {
+		return p, fmt.Errorf("'%v': %w", id, ErrProposalNotFound)
+	}
+
 	return p, err
 }
 
@@ -64,6 +71,10 @@ func (ps *Proposals) GetByReference(ctx context.Context, ref string) (entities.P
 	var p entities.Proposal
 	query := `SELECT * FROM proposals_current WHERE reference=$1 LIMIT 1`
 	err := pgxscan.Get(ctx, ps.Connection, &p, query, ref)
+	if pgxscan.NotFound(err) {
+		return p, fmt.Errorf("'%v': %w", ref, ErrProposalNotFound)
+	}
+
 	return p, err
 }
 

